lib/src/theme/color_extensions.dart

```diff
@@ -0,0 +1,283 @@
+  /// darker, the extension just returns white or black for such attempts, with
+  /// a quick exist from the call.
+  Color getShadeColor({
+    int shadeValue = 15,
+    bool lighten = true,
+    bool keepBlack = true,
+    bool keepWhite = true,
+  }) {
+    if (shadeValue <= 0) return this;
+    int usedShadeValue = shadeValue;
+    if (usedShadeValue > 100) usedShadeValue = 100;
+
+    // Trying to make black darker, just return black
+    if (this == Colors.black && !lighten) return this;
+    // Black is defined to be kept as black.
+    if (this == Colors.black && keepBlack) return this;
+    // Make black lighter as lighten was set and we do not keepBlack
+    if (this == Colors.black) return this.lighten(usedShadeValue);
+
+    // Trying to make white lighter, just return white
+    if (this == Colors.white && lighten) return this;
+    // White is defined to be kept as white.
+    if (this == Colors.white && keepWhite) return this;
+    // Make white darker as we do not keep white.
+    if (this == Colors.white) return darken(usedShadeValue);
+    // We are dealing with some other color than white or black, so we
+    // make it lighter or darker based on flag and requested shade %
+    if (lighten) {
+      return this.lighten(usedShadeValue);
+    } else {
+      return darken(usedShadeValue);
+    }
+  }
+
+  /// Return uppercase Flutter style hex code string of the color.
+  String get hexCode {
+    return value.toRadixString(16).toUpperCase().padLeft(8, '0');
+  }
+
+  /// Applies lightness percentage to color.
+  Color withLightness(double percentage) {
+    final HSLColor hslColor = HSLColor.fromColor(this);
+
+    return hslColor.withLightness(percentage).toColor();
+  }
+
+  /// Gets lightness of color.
+  double get lightness => HSLColor.fromColor(this).lightness;
+
+  /// Calculates the contrast ratio between the current color and the given color [b].
+  ///
+  /// The contrast ratio is computed based on the relative luminance of the two colors
+  /// and is useful for ensuring readability and accessibility of text on backgrounds.
+  ///
+  /// Returns the computed contrast ratio.
+  double contrastRatio(Color b) {
+    final luminanceA = computeLuminance();
+    final luminanceB = b.computeLuminance();
+    if (luminanceA > luminanceB) {
+      return (luminanceA + 0.05) / (luminanceB + 0.05);
+    }
+    return (luminanceB + 0.05) / (luminanceA + 0.05);
+  }
+
+  /// Adjusts the color to a specific contrast target.
+  ///
+  /// Parameters:
+  /// - [targetContrast] The target contrast for the color adjustment.
+  /// - [on] The color to compare contrast against. Defaults to [Colors.white].
+  /// - [maxIterations] The maximum number of iterations for color adjustment. Defaults to 1000 iterations.
+  /// - [epsilon] The difference tolerance for the contrast ratio. Defaults to 0.1.
+  ///
+  /// Returns:
+  /// - A new color with adjusted contrast close to the target contrast.
+  Color adjustContrast(
```

Comment:
Can some of these functions be private?

Reply:
I intentionally kept this method open. It could be beneficial outside of Zeta, especially when we aim to ensure accessibility beyond our standard levels (aa, aaa).